### PR TITLE
refactor(component): component with typescript.

### PR DESCRIPTION
**Pull request by colmugx**
- fix(cli): license
- refactor(component): with typescript

And need help with `factoryComponent`: `props types`.

### CODE REVIEW

File: packages/@remax/components/package.json [hunk 1]
```diff
@@ -4,13 +4,13 @@
   "description": "",
   "main": "src/index.js",
   "scripts": {
-    "watch": "babel src -d lib -w",
-    "build": "babel src -d lib"
+    "watch": "./node_modules/.bin/tsc -w",
```
Comment (aladdin-add) [inline packages/@remax/components/package.json]: nits: 可以直接写成 "tsc -w", 因为 npm scripts 会自动加上 node_modules/

**Comment by CodeFalling** [comment]
@colmugx Thanks for your awesome work ❤️ 

There a two problems now:

### @types/react

I clone this branch run 

```
npm run bootstrap
npm run build
```

got failed, because `tsc` could not find types for `react` module, we need 

```
yarn add --dev @types/react
```

to add dev deps in project root.

### export default vs export

You replace lines of `export` with one `export default`, however, we recommend use 

```js
import {
  Button,
} from '@remax/component';
```

instead of 

```
import RemaxComponent from '@remax/component';
const { Button } = RemaxComponent;
```

Only `export default` will broke other demo now:

image

On the other hand, generate components on runtime will lose autocomplete in Editor & IDE for ts.

File: packages/@remax/components/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  [s: string]: any
+}
+const camelCased = (_str: string) => {
+  const str = _str.split('').map((ch, index) => (index === 0 ? ch.toUpperCase() : ch)).join('');
+  return str.replace(/-([a-z0-9])/g, (g) => { return g[1].toUpperCase(); });
+};
+
+const styleString = (style: CSSProperties) => (
+  Object.entries(style).reduce((styleString, [propName, propValue]) => {
+    return `${styleString}${propName}:${propValue};`;
```
Comment (HuiguangChu) [inline packages/@remax/components/src/index.ts]: remove ``里面的；

File: packages/@remax/components/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  }, '')
+);
+
+function factoryTag(component: string, props: Props<any>) {
+  const Tag = `mini-${component}+${Object.keys(props).filter(propKey => propKey !== 'children').sort().join('+')}`;
```
Comment (HuiguangChu) [inline packages/@remax/components/src/index.ts]: 1. 命名规范：change Tag to tag
2. 冗余变量：remove tag, 直接return  `mini-${component}+${Object.keys(props).filter(propKey => propKey !== 'children').sort().join('+')}`;
3. 不属于Pr问题 @CodeFalling  这个tag将生成“mini-test+【key】”?  为什么不用‘-’连接呢？

File: packages/@remax/components/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+function factoryComponent(component: TComponent) {
+  // props 类型存在问题
+  return <T>(props: Props<T> & IProps) => {
+    const {
+      children
+    } = props;
+    const newProps: IProps = {};
+    for (const propKey of Object.keys(props)) {
+      if (propKey === 'style') {
+        (<string>newProps.style) = styleString(props.style!);
```
Comment (HuiguangChu) [inline packages/@remax/components/src/index.ts]: 这个是不是可以写成 newPros = Object.assign({}, {style : styleString(props.style!)})?

File: packages/@remax/components/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+      } else if (propKey === 'key') {
+        // pass
+      } else {
+        newProps[propKey] = props[propKey];
+      }
+    }
+    const Tag = factoryTag(component, newProps);
+    return React.createElement(Tag, newProps, children);
```
Comment (HuiguangChu) [inline packages/@remax/components/src/index.ts]: return React.createElement(Tag, newProps, ...children);

File: packages/@remax/components/tsconfig.json [hunk 6]
```diff
@@ -0,0 +1,14 @@
+{
+  "compilerOptions": {
+    "module": "commonjs",
```
Comment (HuiguangChu) [inline packages/@remax/components/tsconfig.json]: 这里为什么是common js?  esnext 会不会更好一些呢？

File: packages/@remax/components/tsconfig.json [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+    "target": "es5",
+    "lib": ["es2015", "es2017.object"],
+    "strict": true,
+    "noImplicitAny": true,
+    "esModuleInterop": true,
+    "moduleResolution": "node",
+    "outDir": "lib",
+    "baseUrl": "."
```
Comment (HuiguangChu) [inline packages/@remax/components/tsconfig.json]: "sourceMap": true,
"allowSyntheticDefaultImports": true,

File: packages/@remax/components/tsconfig.json [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+  },
+  "include": ["src/**/*"]
+}
```
Comment (HuiguangChu) [inline packages/@remax/components/tsconfig.json]: "exclude": [
    "node_modules"
  ]

File: packages/@remax/components/src/index.ts [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    for (const propKey of Object.keys(props)) {
+      if (propKey === 'style') {
+        (<string>newProps.style) = styleString(props.style!);
+      } else if (propKey === 'key') {
+        // pass
+      } else {
+        newProps[propKey] = props[propKey];
+      }
+    }
+    const Tag = factoryTag(component, newProps);
```
Comment (HuiguangChu) [inline packages/@remax/components/src/index.ts]: chaneg Tag to tag.
还有就是return 上一行，不需要空行么？

File: packages/@remax/components/src/index.ts (hunk 3 above)
Comment (CodeFalling) [reply packages/@remax/components/src/index.ts]: 为了不和 `mini-test` 的 `-` 混在一起，其实倒是没什么影响

**Comment by CodeFalling** [comment]
一些小问题的修复在：https://github.com/CodeFalling/remax/pull/10

**Pull request merged**